### PR TITLE
Test colls with new generators

### PR DESCRIPTION
**Pull request by gostkin**

### CODE REVIEW

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 1]
```diff
@@ -30,6 +36,7 @@ trait RTypeGens {
       case 1 =>
         pairTypeGenFinal(itemGen, itemGen)
       case _ =>
+        checkDepth(depth)
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGens.scala]: what if depth == 0?

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 2]
```diff
@@ -1,70 +1,99 @@
...
-      col1.indices.toArray shouldBe col1.toArray.indices.toArray
-//      col1.zip(col2).length shouldBe math.min(col1.length, col2.length)
-// TODO     col1.zip(col2).indices.arr shouldBe col1.arr.zip(col2.arr).indices.toArray
+  val typeGenerationDepth = 5
+  val testConfiguration = new GenConfiguration(maxArrayLength = 10)
+  def valueGen[T](t: RType[T]): Gen[T] = rtypeValueGen(testConfiguration)(t)
+
+  val testMinSuccess = MinSuccessful(100)
+  val typeMinSuccess = MinSuccessful(5)
+  val onceMinSuccess = MinSuccessful(1)
```
Comment (aslesarenko) [inline library/src/test/scala/special/collections/CollsTests.scala]: ugly name, any better variant?

Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: fixed

File: library/src/test/scala/scalan/RTypeGens.scala (hunk 1 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
+class GenConfiguration(
+                        val maxArrayLength: Int = 100,
+                        val byteBorders: (Byte, Byte) = (Byte.MinValue, Byte.MaxValue),
+                        val shortBorders: (Short, Short) = (Short.MinValue, Short.MaxValue),
+                        val intBorders: (Int, Int) = (Int.MinValue, Int.MaxValue),
+                        val longBorders: (Long, Long) = (Long.MinValue, Long.MaxValue),
+                        val charBorders: (Char, Char) = (Char.MinValue, Char.MaxValue),
+                        val floatBorders: (Float, Float) = (Float.MinValue, Float.MaxValue),
+                        val doubleBorders: (Double, Double) = (Double.MinValue, Double.MaxValue)
+                      ) {}
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 4]
```diff
@@ -23,7 +31,7 @@ trait RTypeGens {
   val dataTypeGen = Gen.oneOf[RType[_]](primitiveTypeGen, StringType)
 
   def checkDepth(depth: Int): Unit = {
-    if (depth < 0) {
+    if (depth <= 0) {
       throw new RuntimeException(s"Generation depth can't be less then 0, found ${depth}")
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 5]
```diff
@@ -1,27 +1,23 @@
...
 class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGens { testSuite =>
   import Gen._
   import scalan.RType._
   import special.collection._
   import special.collection.ExtensionMethods._
 
-  val typeGenerationDepth = 5
-  val testConfiguration = new GenConfiguration(maxArrayLength = 10)
+  val typeGenerationDepth = 4
+  val testConfiguration = new GenConfiguration(maxArrayLength = 7)
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: tests work a bit slower in this case, i reverted to previous settings now

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 6]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
     }
   }
 
-  // TODO: rework the following tests for every type, now it is adaptation
   property("Coll.flatMap") {
-    forAll(containerOfN[Coll, Int](3, valueGen(IntType)), valueGen(intCollRtype)) { (zs, col) =>
-      val matrix = zs.map(_ => col)
-      val res = zs.zip(matrix).flatMap(_._2)
-      res.toArray shouldBe zs.toArray.flatMap(_ => col.toArray)
+    def runTest[T](externalCol: Coll[Coll[T]], internalCol: Coll[T])(implicit tC: RType[Coll[T]], t: RType[T]) = {
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: fixed

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 7]
```diff
@@ -1,27 +1,23 @@
...
-  val typeGenerationDepth = 5
-  val testConfiguration = new GenConfiguration(maxArrayLength = 10)
+  val typeGenerationDepth = 4
+  val testConfiguration = new GenConfiguration(maxArrayLength = 7)
   def valueGen[T](t: RType[T]): Gen[T] = rtypeValueGen(testConfiguration)(t)
 
   val testMinSuccess = MinSuccessful(100)
   val typeMinSuccess = MinSuccessful(5)
-  val onceMinSuccess = MinSuccessful(1)
+  val oneTimeRunMinSuccess = MinSuccessful(1)
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: fixed

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 8]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
...
+      forAll(valueGen(t), typeMinSuccess) { arr =>
+        whenever(arr.length > 3) {
+          val col = arr(0)
+          val zsArr = arr.slice(1, arr.length)
+          val arrayItemType = t.tA
+          val collItemType = tItem(arrayItemType)
+          val zs = builder.fromArray(zsArr)(arrayItemType)
+          runTest(zs.asInstanceOf[Coll[Coll[Any]]], col.asInstanceOf[Coll[Any]])(arrayItemType.asInstanceOf[RType[Coll[Any]]], collItemType.asInstanceOf[RType[Any]])
+        }
+      }
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: this test checks `flatMap`. the main logic was taken from previous implementation. i just make the test more general

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 9]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
...
     }
   }
 
   property("Coll.segmentLength") {
-    forAll(valueGen(intCollRtype), indexGen, testMinSuccess) { (col: Coll[Int], from: Int) =>
-      col.segmentLength(lt0, from) shouldBe col.toArray.segmentLength(lt0, from)
+    forAll(extendedCollTypeGen(typeGenerationDepth), testMinSuccess) { t: RType[Coll[_]] =>
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: renamed to collTypeGen

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 10]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
...
+      forAll(valueGen(t), indexGen, typeMinSuccess) { (extendedColl, from) =>
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: fixed

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 11]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
...
+          val arrayItemType = t.tA
+          val collItemType = tItem(arrayItemType)
+          val zs = builder.fromArray(zsArr)(arrayItemType)
+          runTest(zs.asInstanceOf[Coll[Coll[Any]]], col.asInstanceOf[Coll[Any]])(arrayItemType.asInstanceOf[RType[Coll[Any]]], collItemType.asInstanceOf[RType[Any]])
+        }
+      }
     }
   }
 
   property("Coll.segmentLength") {
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: we have such operation in collections

File: library/src/test/scala/special/collections/CollsTests.scala [hunk 12]
```diff
@@ -47,55 +54,109 @@ class CollsTests extends PropSpec with PropertyChecks with Matchers with CollGen
...
-//      col.lastIndexWhere(eq0, end) shouldBe col.lastIndexWhere(eq0, end)
-//      def p2(ab: (Int, Int)) = eq0(ab._1) && eq0(ab._2)
-//      col.zip(col).lastIndexWhere(p2, end) shouldBe col.toArray.zip(col.toArray).lastIndexWhere(p2, end)
-//    }
-//  }
-/*
+
+  property("Coll.indexOf") {
+    def runTest[T](usualCol: Coll[T], from: Int)(implicit t: RType[T]) = {
+      val colItem = RTypeUtil.clone(usualCol(0))(usualCol.tItem)
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollsTests.scala]: Removed

File: library/src/test/scala/special/collections/CollGens.scala [hunk 13]
```diff
@@ -4,87 +4,22 @@ import scala.collection.mutable.ArrayBuffer
...
 
-//  val superGenInt = getSuperGen(1, Gen.oneOf(collOverArrayGen, replCollGen, lazyCollGen, lazyUnFuncCollGen))
-//  val superGenByte = getSuperGen(1, Gen.oneOf(bytesOverArrayGen, replBytesCollGen, lazyByteGen))
-//  val superGen = Gen.oneOf(superGenInt, superGenByte)
-
-//  val allGen = Gen.oneOf(superGen, collGen)
-
-//  implicit val arbColl = Arbitrary(collGen)
-//  implicit val arbBytes = Arbitrary(bytesGen)
+  def hashCodeZeroEnd[T](x: T) = x.hashCode() % 10 == 0
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/special/collections/CollGens.scala]: removed

**Pull request merged**